src/node_file.h

```diff
@@ -70,6 +73,64 @@ class FSReqAfterScope {
   Context::Scope context_scope_;
 };
 
+// A wrapper for a file descriptor that will automatically close the fd when
+// the object is garbage collected
+class FD : public AsyncWrap {
+ public:
+  FD(Environment* env, int fd);
+  virtual ~FD();
+
+  int fd() const { return fd_; }
+  size_t self_size() const override { return sizeof(*this); }
+
+  // Will asynchronously close the FD and return a Promise that will
+  // be resolved once closing is complete.
+  static void Close(const FunctionCallbackInfo<Value>& args);
+
+ private:
+  // Synchronous close that emits a warning
+  inline void Close();
+
+  class CloseReq {
```

Comment:
This should inherit from `ReqWrap<uv_fs_t>`, like e.g. `FSReqWrap` does (or just inherit from the latter to begin with, that may or may not be easier)

Reply:
Yeah, considered that but it makes it a bit more heavyweight than it really needs to be. Shouldn't be a problem tho.